nats/bench_command.go

```diff
@@ -64,37 +113,123 @@ func (c *benchCmd) bench(_ *kingpin.ParseContext) error {
+	log.Printf("Starting benchmark [msgs=%s, msgsize=%s, pubs=%d, subs=%d, js=%v, storage=%s, syncpub=%v, pubbatch=%s, pull=%v, pullbatch=%s, request=%v, reply=%v, noqueue=%v, maxackpending=%s, replicas=%d, nopurge=%v, nodelete=%v]", humanize.Comma(int64(c.numMsg)), humanize.IBytes(uint64(c.msgSize)), c.numPubs, c.numSubs, c.js, c.storage, c.syncPub, humanize.Comma(int64(c.pubBatch)), c.pull, humanize.Comma(int64(c.pullBatch)), c.request, c.reply, c.noQueueGroup, humanize.Comma(int64(c.maxAckPending)), c.replicas, c.noPurge, c.noDeleteStream)
 
-	if c.ack && c.progress {
-		log.Printf("Disabling progress bars in request mode")
-		c.progress = false
+	if c.numPubs == 0 && c.numSubs == 0 {
+		log.Fatalf("You must have at least one publisher or at least one subscriber... try adding --pub 1 and/or --sub 1 to the arguments")
+	}
+
+	if c.js && c.maxAckPending != -1 && c.maxAckPending < c.numMsg && c.numSubs > 0 {
+		log.Printf("WARNING: max acks pending is smaller than the mumber of messages and the subscribers are not set to ack, use --ack or you run the risk of the consumers not receiving all the messages")
 	}
 
 	bm := bench.NewBenchmark("NATS", c.numSubs, c.numPubs)
 
 	startwg := &sync.WaitGroup{}
 	donewg := &sync.WaitGroup{}
 
+	var js nats.JetStreamContext
+
+	if (c.request || c.reply) && c.js {
+		log.Fatalf("Request/Reply mode is not applicable to streams")
+	} else if !c.js {
+		if c.request || c.reply {
+			log.Printf("Benchmark in request/reply mode")
+		}
+		if c.request && c.reply {
+			log.Fatalf("Request/Reply mode error: can not be both a requester and a replier at the same time, please use at least two instances of nats bench to benchmark request/reply")
+		}
+		if c.reply && c.numPubs > 0 && c.numSubs > 0 {
+			log.Fatalf("Request/Reply mode error: can not have a publisher while in --reply mode")
+		}
+	}
+
+	if c.js {
+		// create the stream for the benchmark (and purge it)
+		nc, err := nats.Connect(config.ServerURL(), natsOpts()...)
+		if err != nil {
+			log.Fatalf("NATS connection failed: %s", err)
+			return err
+		}
+
+		js, err = nc.JetStream()
+		if err != nil {
+			log.Fatalf("Couldn't create jetstream context: %v", err)
+		}
+
+		// First delete any prior stream unless no delete
+		if !c.noDeleteStream {
+			log.Printf("Deleting any existing stream")
+			js.DeleteStream(JS_STREAM_NAME)
+		}
+
+		storageType := func() nats.StorageType {
+			switch c.storage {
+			case "file":
+				return nats.FileStorage
+			case "memory":
+				return nats.MemoryStorage
+			default:
+				{
+					log.Printf("Unknown storage type %s, using memory", c.storage)
+					return nats.MemoryStorage
+				}
+			}
+		}()
+
+		js.AddStream(&nats.StreamConfig{Name: JS_STREAM_NAME, Subjects: []string{c.subject}, Retention: nats.LimitsPolicy, MaxConsumers: -1, MaxMsgs: -1, MaxBytes: -1, Discard: nats.DiscardOld, MaxAge: 9223372036854775807, MaxMsgsPerSubject: -1, MaxMsgSize: -1, Storage: storageType, Replicas: c.replicas, Duplicates: time.Second * 2})
```

Comment:
surely all these -1 and maxage of 9223372036854775807 isnt required, the js context would do the right thing?

Reply:
Wanted to make sure that those values (which should be large enough for any kind of benchmark scenario) were explicitly set in the source code.